docs/decisions/0045-updated-vector-store-design.md

```diff
@@ -0,0 +1,810 @@
+---
+# These are optional elements. Feel free to remove any of them.
+status: proposed
+contact: westey-m
+date: 2024-05-01
+deciders: sergeymenshykh, markwallace, rbarreto, dmytrostruk, westey-m, matthewbolanos
+consulted: stephentoub, dluc
+informed: 
+---
+
+# Updated Memory Connector Design
+
+## Context and Problem Statement
+
+Semantic Kernel has a collection of connectors to popular Vector databases e.g. Azure AI Search, Chroma, Milvus, ...
+Each Memory connector implements a memory abstraction defined by Semantic Kernel and allows developers to easily integrate Vector databases into their applications.
+The current abstractions are experimental and the purpose of this ADR is to progress the design of the abstractions so that they can graduate to non experimental status.
+
+### Problems with current design
+
+1. The `IMemoryStore` interface has four responsibilities with different cardinalities.
+2. The `IMemoryStore` interface only supports a fixed schema for data storage, retrieval and search, which limits its usability by customers with existing data sets.
+2. The `IMemoryStore` implementations are opinionated around key encoding / decoding and collection name sanitization, which limits its usability by customers with existing data sets.
+
+Responsibilities:
+
+|Functional Area|Cardinality|Significance to Semantic Kernel|
+|-|-|-|
+|Collection/Index create|An implementation per store type and model|Valuable when building a store and adding data|
+|Collection/Index list/exists/delete|An implementation per store type|Valuable when building a store and adding data|
```

Comment:
We actually don't support read right now for collections, i.e. there is no simple get.  I'll update the line there to clarify.  It should really say `Collection/Index list names, exists and delete`, since the closest we get to read is listing the names of indices.  I've previously updated the interface method name to ListCollectionNames already since it was ambiguous too.

You may ask why we don't have a simple get.  If we had that what you retrieve will essentially be a complex schema specific to that implementation and it would need to be converted to a common type.  I think having a common index definition that captures all of the available functionality would be tricky, but certainly having a cut down version is possible.  That is fine when creating, since anything that isn't in the definition just isn't supported, but on get, we would lose information, and that may not be expected for users.